batch2/batch/semaphore.py

```diff
@@ -23,24 +23,30 @@ def __init__(self, sem, weight):
         await self.sem.acquire(self.weight)
 
     async def __aexit__(self, exc_type, exc, tb):
-        await self.sem.release(self.weight)
+        self.sem.release(self.weight)
 
 
-class WeightedSemaphore:
+class FIFOWeightedSemaphore:
     def __init__(self, value=1):
         self.value = value
-        self.cond = asyncio.Condition()
+        self.queue = []
 
     async def acquire(self, weight):
+        event = asyncio.Event()
+        self.queue.append(event)
+
         while self.value < weight:
-            async with self.cond:
-                await self.cond.wait()
+            event.clear()
+            await event.wait()
+
+        self.queue.remove(event)
         self.value -= weight
 
-    async def release(self, weight):
+    def release(self, weight):
         self.value += weight
-        async with self.cond:
-            self.cond.notify_all()
+        if self.queue:
+            first = self.queue[0]
+            first.set()
```

Comment:
So you want to add new things on the left, but pop them off the right.  To do this efficient, you need to use Python deque: "list-like container with fast appends and pops on either end": https://docs.python.org/2/library/collections.html.

While I think this can be made correct, I agree about the worry that the right element is getting popped off.  How about this:

Push the (weight, event) pair onto the queue.  In release, do the comparison/pop/notify.  acquire now does a single wait with no loop.  Cleaner?

Reply:
I did what you suggested except I found it clearer to append new items to the right and pop them off the left.